distributed/http/worker/prometheus/core.py

```diff
@@ -77,19 +77,22 @@ def collect(self):
 
 
 class PrometheusHandler(RequestHandler):
-    _initialized = False
+    _collector: WorkerMetricCollector | None = None
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, *args, dask_server=None, **kwargs):
         import prometheus_client
 
-        super().__init__(*args, **kwargs)
+        super().__init__(*args, dask_server=dask_server, **kwargs)
 
-        if PrometheusHandler._initialized:
+        if PrometheusHandler._collector:
+            # Especially during testing, multiple workers are started
+            # sequentially in the same python process
+            PrometheusHandler._collector.server = self.server
             return
 
-        prometheus_client.REGISTRY.register(WorkerMetricCollector(self.server))
-
-        PrometheusHandler._initialized = True
+        PrometheusHandler._collector = WorkerMetricCollector(self.server)
+        # Register collector
+        prometheus_client.REGISTRY.register(PrometheusHandler._collector)
```

Comment:
If there are multiple concurrent `Worker` instances in the same process (very common with async workers in all `gen_cluster` tests), this means `PrometheusHandler._collector` will only be set to the `WorkerMetricCollector` of the last worker to start. Is that okay? Does the `prometheus_client.REGISTRY.register(PrometheusHandler._collector)` mean that all the collectors will be registered somewhere anywhere?

I don't know what either `_collector` or `prometheus_client.REGISTRY` are for, I'm just seeing some global variables getting overwritten in a place where I know there might not just be multiple workers sequentially, but also multiple workers in parallel.

Reply:
The way this had been implemented before was that we'd simply set up a `PrometheusHandler` for the first worker spinning up in the process, set `_initialized` to `True` and call it a day. This led to `test_prometheus_collect_task_states` because that `PrometheusHandler` did not belong to the newly spun-up worker when running multiple tests. This new pattern surely isn't perfect as we only ever have a single instance of a `PrometheusHandler` per process talking to the last worker we spun up. However, IMO this is better than what we had before and it would require some more serious thinking to understand the Prometheus client to handle parallel workers gracefully. I'm happy to file a follow-up issue for that.

Regarding the singleton implementation: I've stolen that from https://github.com/hendrikmakait/distributed/blob/13e315c11c7277ba0cadd9f5c2a16364cdeaf14b/distributed/http/scheduler/prometheus/core.py#L79